src/algebra/lie/free.lean

```diff
@@ -0,0 +1,229 @@
+  smul_add  := by { rintros t ⟨a⟩ ⟨b⟩, change quot.mk _ _ = quot.mk _ _, rw smul_add, },
+  zero_smul := by { rintros ⟨a⟩, change quot.mk _ _ = quot.mk _ _, rw zero_smul, },
+  smul_zero := λ t, by { change quot.mk _ _ = quot.mk _ _, rw smul_zero, }, }
+
+instance : lie_algebra R (free_lie_algebra R X) :=
+{ lie_smul :=
+  begin
+    rintros t ⟨a⟩ ⟨c⟩,
+    change quot.mk _ (a • (t • c)) = quot.mk _ (t • (a • c)),
+    rw ← smul_comm,
+  end, }
+
+variables {X}
+
+/-- The embedding of `X` into the free Lie algebra of `X` with coefficients in the commutative ring
+`R`. -/
+def of : X → free_lie_algebra R X := λ x, quot.mk _ (lib.of R x)
+
+variables {L : Type w} [lie_ring L] [lie_algebra R L]
+
+local attribute [instance] lie_ring.to_non_unital_non_assoc_semiring
+local attribute [instance] lie_algebra.is_scalar_tower
+local attribute [instance] lie_algebra.smul_comm_class
+
+/-- An auxiliary definition used to construct the equivalence `lift` below. -/
+def lift_aux (f : X → L) := lib.lift R f
+
+lemma lift_aux_map_smul (f : X → L) (t : R) (a : lib R X) :
+  lift_aux R f (t • a) = t • lift_aux R f a :=
+non_unital_alg_hom.map_smul _ t a
+
+lemma lift_aux_map_add (f : X → L) (a b : lib R X) :
+  lift_aux R f (a + b) = (lift_aux R f a) + (lift_aux R f b) :=
+non_unital_alg_hom.map_add _ a b
+
+lemma lift_aux_map_mul (f : X → L) (a b : lib R X) :
+  lift_aux R f (a * b) = ⁅lift_aux R f a, lift_aux R f b⁆ :=
+non_unital_alg_hom.map_mul _ a b
+
+lemma lift_aux_spec (f : X → L) (a b : lib R X) (h : free_lie_algebra.rel R X a b) :
+  lift_aux R f a = lift_aux R f b :=
+begin
+  induction h with c a' b' c' t a' b' h₁ h₂ a' b' c' h₁ h₂ a' b' c' h₁ h₂ a' b' c' h₁ h₂,
+  { -- rel.lie_self
+    simp only [lift_aux_map_mul, non_unital_alg_hom.map_zero, lie_self], },
+  { -- rel.leibniz_lie
+    simp only [lift_aux_map_mul, lift_aux_map_add, sub_add_cancel, lie_lie], },
+  { -- rel.smul
+    simp only [lift_aux_map_smul, h₂], },
+  { -- rel.add
+    simp only [lift_aux_map_add, h₂], },
+  { -- rel.mul_left
+    simp only [lift_aux_map_mul, h₂], },
+  { -- rel.mul_right
+    simp only [lift_aux_map_mul, h₂], },
+end
+
+/-- The functor `X ↦ free_lie_algebra R X` from the category of types to the category of Lie
+algebras over `R` is adjoint to the forgetful functor in the other direction. -/
+def lift : (X → L) ≃ (free_lie_algebra R X →ₗ⁅R⁆ L) :=
+{ to_fun    := λ f,
+    { to_fun    := λ c, quot.lift_on c (lift_aux R f) (lift_aux_spec R f),
+      map_add'  := by { rintros ⟨a⟩ ⟨b⟩, rw ← lift_aux_map_add, refl, },
+      map_smul' := by { rintros t ⟨a⟩, rw ← lift_aux_map_smul, refl, },
+      map_lie'  := by { rintros ⟨a⟩ ⟨b⟩, rw ← lift_aux_map_mul, refl, }, },
+  inv_fun   := λ F, F ∘ (of R),
+  left_inv  := λ f, by { ext x, simp only [lift_aux, of, quot.lift_on_mk, lie_hom.coe_mk,
+    function.comp_app, lib.lift_of_apply], },
+  right_inv := λ F,
+    begin
+      let F' : non_unital_alg_hom R (lib R X) L :=
+      { to_fun    := F ∘ (quot.mk (rel R X)),
```

Comment:
Can this be expressed as `F.comp ...` somehow?

Reply:
My suspicion is that you can define a `to_non_unital_alg_hom` field on `→ₗ⁅R⁆` in the presence of the ring structure.

Reply:
I'll try again now!

Reply:
Got pulled into something else so just got to this. It works. Will push shortly.